src/components/Attachment/hooks/useAudioController.ts

```diff
@@ -1,17 +1,61 @@
 import { useCallback, useEffect, useRef, useState } from 'react';
+import { useChannelActionContext, useTranslationContext } from '../../../context';
+
+export const elementIsPlaying = (audioElement: HTMLAudioElement | null) =>
+  audioElement && !(audioElement.paused || audioElement.ended);
 
 export const PROGRESS_UPDATE_INTERVAL = 100;
 
-export const useAudioController = () => {
+const DEFAULT_PLAYBACK_RATES = [1.0, 1.5, 2.0];
+
+type AudioControllerParams = {
+  /** Audio duration in seconds. */
+  durationSeconds?: number;
+  /** An array of fractional numeric values of playback speed to override the defaults (1.0, 1.5, 2.0) */
+  playbackRates?: number[];
+};
+
+export const useAudioController = (params: AudioControllerParams = {}) => {
+  const { addNotification } = useChannelActionContext('useAudioController');
+  const { t } = useTranslationContext('useAudioController');
+  const { durationSeconds } = params;
+  const playbackRates = params.playbackRates || DEFAULT_PLAYBACK_RATES;
   const [isPlaying, setIsPlaying] = useState(false);
   const [progress, setProgress] = useState(0);
```

Comment:
Can `progress` be just a variable computed based on `secondsElapsed` and `durationSeconds`? I think it might make the logic a bit more straightforward.

Reply:
Remove the `progress` state variable and calculate the progress value on return from the hook.

https://github.com/GetStream/stream-chat-react/pull/2311/commits/ccfd5b91f2ace839e2979c290615194e4d13caf2